src/dialog/unix/SDL_zenitydialog.c

```diff
@@ -348,49 +304,40 @@ static void run_zenity(zenityArgs* arg_struct)
     }
 }
 
-static int run_zenity_thread(void* ptr)
+static void free_zenity_args(zenityArgs *args)
+{
+    for (int i = 0; i < args->nfilters; i++) {
+        SDL_free(args->filters_slice[i]);
+    }
+    SDL_free(args->filename);
+    SDL_free(args->title);
+    SDL_free(args->accept);
+    SDL_free(args->cancel);
+    SDL_free(args->argv);
+    SDL_free(args);
+}
+
+static int run_zenity_thread(void *ptr)
 {
-    run_zenity(ptr);
-    SDL_free(ptr);
+    zenityArgs *args = ptr;
+    run_zenity(args->callback, args->userdata, args->argv);
+    free_zenity_args(args);
     return 0;
 }
 
 void SDL_Zenity_ShowFileDialogWithProperties(SDL_FileDialogType type, SDL_DialogFileCallback callback, void *userdata, SDL_PropertiesID props)
 {
-    zenityArgs *args;
-    SDL_Thread *thread;
+    zenityArgs *args = SDL_malloc(sizeof(*args));
 
-    args = SDL_malloc(sizeof(*args));
-    if (!args) {
+    if (!args || !generate_zenity_args(type, callback, userdata, props, args)) {
         callback(userdata, NULL, -1);
```

Comment:
If `args` is allocated and `generate_zenity_args` fails, does `args` need to be freed?